### PR TITLE
Deprecate instructions argument in class Schedule

### LINKED ISSUE #4872: Deprecate initializing Schedules with Instructions
### What is the expected enhancement?
It is currently possible to initialize a Schedule with timed instructions like `Schedule((100, Play(Constant(100, 1.0), DriveChannel(0))))`. This should be deprecated in the next release so as to enable us to add signature information to schedules in future releases.

### PR DESCRIPTION
**Pull request by mlarocca**
### Summary

Deprecate initializing `Schedule`s with Instructions so as to enable adding signature information to schedules in future releases.

### Details and comments

- Adds a warning when *schedules parameters are passed to `Schedule`'s constructor as well as in `ParameterizedSchedule`'s constructor.
- Remove the usage of the deprecated arguments in schedule.py
- Remove the usage of the deprecated arguments in instruction.py
- Mark the usage of the deprecated arguments in all other modules and tests.
- Add a release note about this deprecation.

-----

**Work in progress**

There are a few questions to be answered, about:
* The best way to deprecate arguments
* If the usage of said deprecated  arguments should be cleaned up, marked as ToDo, or neither.
* Release notes.

Please see comments to the files

resolves #4872

### CODE REVIEW

File: qiskit/assembler/assemble_schedules.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def _assemble_experiments(
     instruction_converter = instruction_converter(qobj.PulseQobjInstruction,
                                                   **run_config.to_dict())
 
+    #ToDo(4872) remove the use of deprecated constructor Schedule(*schedulers)
```
Comment (mlarocca) [inline qiskit/assembler/assemble_schedules.py:95]: I marked all the calls to `Schedule` that use the deprecated argument.
Should we:
- Remove the use of the deprecated argument
- Leave the ToDo (possible updating the issue number with a newly created issue to remove said usage)
- Do nothing, remove all these comments.

Which option do you think is the best? Or would you like to handle these in yet a different way?

File: qiskit/pulse/instructions/instruction.py [hunk 2]
```diff
@@ -190,7 +190,9 @@ def shift(self: ScheduleComponent, time: int, name: Optional[str] = None) -> Sch
         """
         if name is None:
             name = self.name
-        return Schedule((time, self), name=name)
```
Comment (mlarocca) [inline qiskit/pulse/instructions/instruction.py:193]: Since `Instruction` is also a subclass of `ScheduleComponent`, I went ahead and removed the usage of the deprecated argument.
I can obviously roll-back or handle this according to your guidelines.

File: qiskit/pulse/schedule.py [hunk 3]
```diff
@@ -64,13 +66,18 @@ def __init__(self, *schedules: List[Union[ScheduleComponent, Tuple[int, Schedule
...
 
-        for sched_pair in schedules:
-            try:
-                time, sched = sched_pair
-            except TypeError:
-                # recreate as sequence starting at 0.
-                time, sched = 0, sched_pair
-            self._mutable_insert(time, sched)
+        if len(schedules) > 0:
+            warnings.warn('Argument schedules for the constructor of class qiskit.pulse.Schedule '
```
Comment (mlarocca) [inline qiskit/pulse/schedule.py:70]: I wasn't sure what was the best way to deprecate an argument. I saw that often it's used qiskit.util.deprecate_arguments, but that seems tailored for the situation where you need to rename a named argument, isn't it?

Please feel free to let me know if there is a better way, or if you'd like to add a new decorator to qiskit.util to handle this situation.

File: qiskit/pulse/schedule.py [hunk 4]
```diff
@@ -203,7 +210,10 @@ def _immutable_shift(self,
         """
         if name is None:
             name = self.name
-        return Schedule((time, self), name=name)
```
Comment (mlarocca) [inline qiskit/pulse/schedule.py:206]: I removed all usage of the deprecated constructor in this file.

File: qiskit/pulse/schedule.py [hunk 5]
```diff
@@ -769,17 +789,24 @@ def __init__(self, *schedules, parameters: Optional[Dict[str, Union[float, compl
...
-            if isinstance(schedule, ParameterizedSchedule):
-                parameterized.append(schedule)
-                parameters += schedule.parameters
-            elif callable(schedule):
-                parameterized.append(schedule)
-            elif isinstance(schedule, Schedule):
-                full_schedules.append(schedule)
-            else:
-                raise PulseError('Input type: {} not supported'.format(type(schedule)))
+
```
Comment (mlarocca) [inline qiskit/pulse/schedule.py:792]: I wasn't sure about this, but I supposed you wanted to deprecate the argument for `ParameterizedSchedule` too. Just let me know if that's not the case, and I'll revert the change.

File: releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (mlarocca) [inline releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml:1]: I couldn't understand it, where do you take the SHA appended to the releasenotes' filenames?

File: qiskit/pulse/instructions/instruction.py [hunk 7]
```diff
@@ -204,7 +206,10 @@ def insert(self, start_time: int, schedule: ScheduleComponent,
         """
         if name is None:
             name = self.name
-        return Schedule(self, (start_time, schedule), name=name)
+        new_schedule = Schedule(name=name)
+        new_schedule.insert(0, self, inplace=True)
+        new_schedule.insert(start_time, schedule, inplace=True)
```
Comment (mlarocca) [inline qiskit/pulse/instructions/instruction.py:211]: As far as I could understand, this was the best way to have "`schedule` inserted within`self` at `start_time`"... is that right? (I double checked it's covered by having tests fail before by removing the second line)

File: qiskit/pulse/schedule.py [hunk 8]
```diff
@@ -309,7 +319,10 @@ def append(self, schedule: ScheduleComponent,
 
     def flatten(self) -> 'Schedule':
         """Return a new schedule which is the flattened schedule contained all ``instructions``."""
-        return Schedule(*self.instructions, name=self.name)
+        schedule = Schedule(name=self.name)
+        for instruction in self.instructions:
+            schedule.insert(*instruction, inplace=True)
```
Comment (mlarocca) [inline qiskit/pulse/schedule.py:324]: Considering the constructor is being deprecated, should a new method that insert a list of schedules be added, to avoid having to use this pattern in several places? I counted 3 occurrences in this file, and at least that many (probably more) in the rest of the library.

Comment (taalexander) [reply qiskit/pulse/schedule.py:324]: I believe this is better suited by a for loop:
```python
for t0, inst in instructions:
     self.insert(t0, instruction, inplace=True)
```
Outside of this class, the functionality is not often used.

File: releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml (hunk 6 above)
Comment (taalexander) [reply releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml:1]: DId you use the reno CLI tool?
```
reno new 4872-deprecate-...
```

File: qiskit/pulse/schedule.py [hunk 9]
```diff
@@ -64,13 +66,18 @@ def __init__(self, *schedules: List[Union[ScheduleComponent, Tuple[int, Schedule
...
-        for sched_pair in schedules:
-            try:
-                time, sched = sched_pair
-            except TypeError:
-                # recreate as sequence starting at 0.
-                time, sched = 0, sched_pair
-            self._mutable_insert(time, sched)
+        if len(schedules) > 0:
+            warnings.warn('Argument schedules for the constructor of class qiskit.pulse.Schedule '
+                          'is deprecated as of 0.22.0, and will be removed in a later release.',
```
Comment (taalexander) [inline qiskit/pulse/schedule.py:71]:
```suggestion
            warnings.warn(
                          'Constructing a Schedule with Instructions/Schedules '
                          'is deprecated as of 0.22.0, and will be removed in a later release.',
```

File: qiskit/pulse/schedule.py [hunk 10]
```diff
@@ -769,17 +789,24 @@ def __init__(self, *schedules, parameters: Optional[Dict[str, Union[float, compl
...
-                parameters += schedule.parameters
-            elif callable(schedule):
-                parameterized.append(schedule)
-            elif isinstance(schedule, Schedule):
-                full_schedules.append(schedule)
-            else:
-                raise PulseError('Input type: {} not supported'.format(type(schedule)))
+
+        if len(schedules) > 0:
+            warnings.warn('Argument schedules for the constructor of class '
```
Comment (taalexander) [inline qiskit/pulse/schedule.py:794]: Please update this message as well 😄

File: releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+deprecations:
+  - |
+    Argument `schedule` for :class:`qiskit.pulse.Schedule` has been
+    deprecated and will be removed in a later release.
+    You can still use `Schedule()` or `Schedule(name=)`.
```
Comment (taalexander) [inline releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml:6]: Similar to the deprecation warnings. `Constructing :class:`qiskit.pulse.Schedule`s with ...

File: test/python/pulse/test_schedule.py [hunk 12]
```diff
@@ -946,18 +958,21 @@ def test_single_channel_out_of_order(self):
...
         instructions = [(0, ShiftPhase(0, DriveChannel(1))),
                         (1, Acquire(10, AcquireChannel(0), MemorySlot(1)))]
 
+        # ToDo(4872) remove the use of deprecated constructor Schedule(*schedules)
         self.assertEqual(Schedule(*instructions), Schedule(*reversed(instructions)))
 
     def test_different_name_equal(self):
         """Test that names are ignored when checking equality."""
 
+        # ToDo(4872) remove the use of deprecated constructor Schedule(*schedules)
```
Comment (taalexander) [inline test/python/pulse/test_schedule.py:975]: These changes will certainly make code like this more verbose 😦 . After working through these changes what do you think about them as a whole @mlarocca?

**Review by taalexander — CHANGES_REQUESTED** [review]
These changes will certainly make code like this more verbose 😦 . After working through these changes what do you think about them as a whole @mlarocca?

File: qiskit/pulse/schedule.py (hunk 8 above)
Comment (mlarocca) [reply qiskit/pulse/schedule.py:325]: What do you think is best? It's no bother for me to add the new method, but if it's not necessary/valuable, we can leave it out, of course.

File: releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml (hunk 6 above)
Comment (mlarocca) [reply releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml:1]: Ah, I had missed that! Thanks :-)

Comment (mlarocca) [reply releasenotes/notes/4872-Deprecate-Schedule-input-arg-schedules.yaml]: Ah, I had missed that! Thanks :-)

File: test/python/pulse/test_schedule.py (hunk 12 above)
Comment (mlarocca) [reply test/python/pulse/test_schedule.py:987]: Overall I think there were a dozen places where this had to be changed (tests excluded), in various ways.
It's a bit more verbose, of course, but not that bad.
The one thing that worries me a bit is that right now refactoring the code to remove the deprecated constructor is a bit tricky, because if you forget to use `inplace=True` named argument on `.insert`, you mess up - if you have tests then you are golden, you'll notice, but client code might not be as well covered... you know.

So maybe it makes sense to add the `insert_iter` method, as you had suggested?

File: qiskit/qobj/converters/pulse_instruction.py [hunk 13]
```diff
@@ -508,6 +508,7 @@ def gen_fc_sched(*args, **kwargs):
                 _phase = phase_expr(*args, **kwargs)
                 return instructions.SetPhase(_phase, channel) << t0
 
+            # ToDo(4872) remove the use of deprecated constructor Schedule(*schedulers)
```
Comment (mlarocca) [inline qiskit/qobj/converters/pulse_instruction.py:511]: So, I noticed these are not covered in tests (so it wasn't failing/generate warnings).
I still have to find the best way to replace this, I couldn't manage to have e a proper look at `ParameterizedSchedule` today, but at first glance it seems we might need a new method to insert a "callable"?

File: qiskit/assembler/assemble_schedules.py [hunk 14]
```diff
@@ -92,9 +93,18 @@ def _assemble_experiments(
     instruction_converter = instruction_converter(qobj.PulseQobjInstruction,
                                                   **run_config.to_dict())
 
-    schedules = [
-        sched if isinstance(sched, pulse.Schedule) else pulse.Schedule(sched) for sched in schedules
```
Comment (mlarocca) [inline qiskit/assembler/assemble_schedules.py:96]: This was a bit tricky to figure out, also because in the mypi signature it reads:
`schedules: List[pulse.Schedule]`

I suppose, also looking at tests, that should better be `List[pulse.ScheduleComponent]`?

Is it possible to have other types, like (time, instruction) tuples?

File: qiskit/assembler/assemble_schedules.py [hunk 15]
```diff
@@ -92,9 +93,18 @@ def _assemble_experiments(
...
+    def schedule_converter(sched):
+        if isinstance(sched, pulse.Schedule):
+            return sched
+        elif isinstance(sched, ScheduleComponent):
+            new_sched = pulse.Schedule()
+            for t0, instr in sched.instructions:
+                new_sched.insert(t0, instr, inplace=True)
+            return new_sched
+        else:
+            raise QiskitError('Invalid parameter schedules: {}.'.format(type(sched)))
```
Comment (mlarocca) [inline qiskit/assembler/assemble_schedules.py:105]: Should we bother checking or can we assume (as it was doing before) that's either a  `Schedule` or a `ScheduleComponent`?

**Review by mlarocca — COMMENTED** [review]
@taalexander thanks for your feedback!
I tried to implement all the suggestions. Just have a couple of new questions that came up in the process.

File: qiskit/qobj/converters/pulse_instruction.py [hunk 16]
```diff
@@ -558,6 +560,7 @@ def gen_sf_schedule(*args, **kwargs):
                 _frequency = frequency_expr(*args, **kwargs)
                 return instructions.SetFrequency(_frequency, channel) << t0
 
+            # ToDo(4872) remove the use of deprecated constructor Schedule(*schedulers)
```
Comment (mlarocca) [inline qiskit/qobj/converters/pulse_instruction.py:563]: I think there was a bug here:
https://github.com/Qiskit/qiskit-terra/pull/5093/files#diff-78f2c68464fc91f622184a3afeaf0a8bR554

https://github.com/Qiskit/qiskit-terra/pull/5093/files#diff-78f2c68464fc91f622184a3afeaf0a8bR580

While adding tests I noticed that in case frequency was passed as a string, it still tries to multiply it by 1e9.
I'll add a fix for these as well.

File: qiskit/qobj/converters/pulse_instruction.py [hunk 17]
```diff
@@ -544,21 +550,27 @@ def convert_set_frequency(self, instruction):
...
-                return instructions.SetFrequency(_frequency, channel) << t0
+                return instructions.SetFrequency(_frequency * FREQUENCY_SCALING_FACTOR,
+                                                 channel) << t0
 
-            return ParameterizedSchedule(gen_sf_schedule, parameters=frequency_expr.params)
+            schedule = ParameterizedSchedule(parameters=frequency_expr.params)
+            schedule.add_parameterized(gen_sf_schedule)
+            return schedule
+        else:
+            frequency = frequency * FREQUENCY_SCALING_FACTOR
```
Comment (mlarocca) [inline qiskit/qobj/converters/pulse_instruction.py:573]: @taalexander I'm a bit worried here about this happening under the hood... I added something in the docs and a constant, to try and make this clearer and more consistent, but let me know if you'd like a different solution.

File: qiskit/qobj/converters/pulse_instruction.py [hunk 18]
```diff
@@ -544,21 +550,27 @@ def convert_set_frequency(self, instruction):
 
         Args:
             instruction (PulseQobjInstruction): set frequency qobj instruction
+                                                The passed frequency will be scaled by a factor 1e9
```
Comment (mlarocca) [inline qiskit/qobj/converters/pulse_instruction.py:553]: On second thought, maybe it would be better to say something like: "Frequency is intended in (GHz)"?

File: qiskit/qobj/converters/pulse_instruction.py (hunk 17 above)
Comment (taalexander) [reply qiskit/qobj/converters/pulse_instruction.py:573]: This is a result of Qiskit using Si units and the backend using GHz, the format of each is defined in the schema/the Qiskit objects.

File: qiskit/qobj/converters/pulse_instruction.py (hunk 16 above)
Comment (taalexander) [reply qiskit/qobj/converters/pulse_instruction.py]: Thank you!

File: qiskit/assembler/assemble_schedules.py (hunk 14 above)
Comment (taalexander) [reply qiskit/assembler/assemble_schedules.py:96]: Yes, it is also possible to get a `(time, instruction)` tuple, this conversion would have been handled by the `pulse.Schedule` in the past.

File: qiskit/qobj/converters/pulse_instruction.py (hunk 13 above)
Comment (taalexander) [reply qiskit/qobj/converters/pulse_instruction.py]: I don't mind leaving the ParameterizedSchedule as is, it will be deprecated by #5073.

File: test/python/pulse/test_schedule.py (hunk 12 above)
Comment (taalexander) [reply test/python/pulse/test_schedule.py:987]: The other option (which may be frustrating :( ) is to leave the current functionality. Why, I had initially proposed removing the instructions int the constructors was to turn `Schedule` into a representation of a function in the pulse IR. I had imagined.
```python
sched = Schedule(signature)
sched.append(Instruction(100, sched.params[0])) 
...
caller = Schedule()
caller.call(sched, args=[1.0])
```
In this way having Schedule be a representation of a function within the IR. However, the way parameters work in circuits and now schedules with #5073 is that the parameters of the object are automatically extended when a new instruction with a parameter is added. This means that having a signature in the constructor would be pointless in these objects as it would be wrong as soon as a new instruction with a parameter would be added.

The workaround to this is to view the `QuantumCircuit`, `Schedule` as basic blocks, that just contain instructions. A `Function` (which doesn't exist currently in Qiskit) should have a signature well defined and be composed of several basic blocks.

The reason I bring this up is that if we are not going to turn `Schedule` into a `Function` with a signature than we don't necessarily have to change the current constructor behaviour. Do you have any thoughts about this?

Comment (mlarocca) [reply test/python/pulse/test_schedule.py:987]: @taalexander 
Thanks again for your feedback! No worries, no frustration - even if we decide to discard these changes, it will have been useful for me to get to know better the code base and how this Schedule/Instruction concept works :)

About your question: my take is that objects for which you pass a signature should indeed be immutable (or otherwise invalidated on `.append`), ideally I would pass both a signature and the instructions/parameters list in the constructor.

So, I agree with you on this, #5073 makes having a signature in the constructor pointless.
I like your idea of going for `Function` and `Schedule` classes. I guess the future `Function` class would be similar to what's currently in schedule.py, did I get that right?

Anyway, bottom line: no problem if you'd like to close this PR and gloss over the changes, I can create a new PR with just the couple of bug fixes and tests that I had added.

How would you like to proceed?
Thanks!

**Comment by lcapelluto** [comment]
This interface change is superseded by the introduction of `ScheduleBlock`, for instance: https://github.com/Qiskit/qiskit-terra/pull/5854
